File: clib/demos/32bit/C/tttrmode/tttrmode.c
/************************************************************************

Demo access to MultiHarp 150 Hardware via MHLIB v.2.0
The program performs a measurement based on hardcoded settings.
The resulting event data is stored in a binary output file.

Michael Wahl, PicoQuant GmbH, May 2020

Note: This is a console application (i.e. run in Windows cmd box)

Note: At the API level the input channel numbers are indexed 0..N-1
where N is the number of input channels the device has.

Note: This demo writes only raw event data to the output file.
It does not write a file header as regular .ht* files have it.


Tested with the following compilers:

  - MinGW 2.0.0 (Windows 32 bit)
  - MinGW-W64 4.3.5 (Windows 64 bit)
  - MS Visual C++ 6.0 (Windows 32 bit)
  - MS Visual C++ 2015 and 2019 (Windows 32 and 64 bit)
  - gcc 4.8.3 and 7.5.0 (64 bit)

************************************************************************/

#ifndef _WIN32
#include <unistd.h>
#define Sleep(msec) usleep(msec*1000)
#define __int64 long long
#else
#include <windows.h>
#include <dos.h>
#include <conio.h>
#endif

#include <stdio.h>
#include <stdlib.h>
#include <string.h>

#include "mhdefin.h"
#include "mhlib.h"
#include "errorcodes.h"


unsigned int buffer[TTREADMAX];


int main(int argc, char* argv[])
{

  int dev[MAXDEVNUM];
  int found = 0;
  FILE *fpout;
  int retcode;
  int ctcstatus;
  char LIB_Version[8];
  char HW_Model[32];
  char HW_Partno[8];
  char HW_Serial[9];
  char HW_Version[16];
  char Errorstring[40];
  int NumChannels;
  int Mode = MODE_T2; //set T2 or T3 here, observe suitable Sync divider and Range!
  int Binning = 0; //you can change this, meaningful only in T3 mode
  int Offset = 0;  //you can change this, meaningful only in T3 mode
  int Tacq = 10000; //Measurement time in millisec, you can change this
  int SyncDivider = 1; //you can change this, observe Mode! READ MANUAL!

  int SyncTiggerEdge = 0; //you can change this
  int SyncTriggerLevel = -50; //you can change this
  int InputTriggerEdge = 0; //you can change this
  int InputTriggerLevel = -50; //you can change this

  double Resolution;
  int Syncrate;
  int Countrate;
  int i;
  int flags;
  int warnings;
  char warningstext[16384]; //must have 16384 bytest text buffer
  int nRecords;
  unsigned int Progress;


  printf("\nMultiHarp 150 MHLib.DLL Demo Application              PicoQuant GmbH, 2020");
  printf("\n~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~");
  MH_GetLibraryVersion(LIB_Version);
  printf("\nLibrary version is %s\n", LIB_Version);
  if (strncmp(LIB_Version, LIB_VERSION, sizeof(LIB_VERSION)) != 0)
  {
    printf("\nWarning: The application was built for version %s.", LIB_VERSION);
  }

  if ((fpout = fopen("tttrmode.out", "wb")) == NULL)
  {
    printf("\ncannot open output file\n");
    goto ex;
  }


  printf("\nSearching for MultiHarp devices...");
  printf("\nDevidx     Serial     Status");


  for (i = 0; i < MAXDEVNUM; i++)
  {
    retcode = MH_OpenDevice(i, HW_Serial);
    if (retcode == 0) //Grab any device we can open
    {
      printf("\n  %1d        %7s    open ok", i, HW_Serial);
      dev[found] = i; //keep index to devices we want to use
      found++;
    }
    else
    {
      if (retcode == MH_ERROR_DEVICE_OPEN_FAIL)
      {
        printf("\n  %1d        %7s    no device", i, HW_Serial);
      }
      else
      {
        MH_GetErrorString(Errorstring, retcode);
        printf("\n  %1d        %7s    %s", i, HW_Serial, Errorstring);
      }
    }
  }

  //In this demo we will use the first device we find, i.e. dev[0].
  //You can also use multiple devices in parallel.
  //You can also check for specific serial numbers, so that you always know 
  //which physical device you are talking to.

  if (found < 1)
  {
    printf("\nNo device available.");
    goto ex;
  }
  printf("\nUsing device #%1d", dev[0]);
  printf("\nInitializing the device...");


  retcode = MH_Initialize(dev[0], Mode, 0);
  if (retcode < 0)
  {
    MH_GetErrorString(Errorstring, retcode);
    printf("\nMH_Initialize error %d (%s). Aborted.\n", retcode, Errorstring);
    goto ex;
  }

  retcode = MH_GetHardwareInfo(dev[0], HW_Model, HW_Partno, HW_Version);
  if (retcode < 0)
  {
    MH_GetErrorString(Errorstring, retcode);
    printf("\nMH_GetHardwareInfo error %d (%s). Aborted.\n", retcode, Errorstring);
    goto ex;
  }
  else
  {
    printf("\nFound Model %s Part no %s Version %s", HW_Model, HW_Partno, HW_Version);
  }


  retcode = MH_GetNumOfInputChannels(dev[0], &NumChannels);
  if (retcode < 0)
  {
    MH_GetErrorString(Errorstring, retcode);
    printf("\nMH_GetNumOfInputChannels error %d (%s). Aborted.\n", retcode, Errorstring);
    goto ex;
  }
  else
  {
    printf("\nDevice has %i input channels.", NumChannels);
  }

  printf("\n\nUsing the following settings:\n");

  printf("Mode              : %d\n", Mode);
  printf("Binning           : %d\n", Binning);
  printf("Offset            : %d\n", Offset);
  printf("AcquisitionTime   : %d\n", Tacq);
  printf("SyncDivider       : %d\n", SyncDivider);
  printf("SyncTiggerEdge    : %d\n", SyncTiggerEdge);
  printf("SyncTriggerLevel  : %d\n", SyncTriggerLevel);
  printf("InputTriggerEdge  : %d\n", InputTriggerEdge);
  printf("InputTriggerLevel : %d\n", InputTriggerLevel);


  retcode = MH_SetSyncDiv(dev[0], SyncDivider);
  if (retcode < 0)
  {
   MH_GetErrorString(Errorstring, retcode);
    printf("\nPH_SetSyncDiv error %d (%s). Aborted.\n", retcode, Errorstring);
    goto ex;
  }

  retcode =MH_SetSyncEdgeTrg(dev[0], SyncTriggerLevel, SyncTiggerEdge);
  if (retcode < 0)
  {
    MH_GetErrorString(Errorstring, retcode);
    printf("\nMH_SetSyncEdgeTrg error %d (%s). Aborted.\n", retcode, Errorstring);
    goto ex;
  }

  retcode = MH_SetSyncChannelOffset(dev[0], 0);
  if (retcode<0)
  {
   MH_GetErrorString(Errorstring, retcode);
    printf("\nMH_SetSyncChannelOffset error %d (%s). Aborted.\n", retcode, Errorstring);
    goto ex;
  }

  for (i = 0; i < NumChannels; i++) // we use the same input offset for all channels
  {
    retcode = MH_SetInputEdgeTrg(dev[0], i, InputTriggerLevel, InputTriggerEdge);
    if (retcode < 0)
    {
      MH_GetErrorString(Errorstring, retcode);
      printf("\nMH_SetInputEdgeTrg error %d (%s). Aborted.\n", retcode, Errorstring);
      goto ex;
    }

    retcode =MH_SetInputChannelOffset(dev[0], i, 0);
    if (retcode < 0)
    {
      MH_GetErrorString(Errorstring, retcode);
      printf("\nMH_SetInputChannelOffset error %d (%s). Aborted.\n", retcode, Errorstring);
      goto ex;
    }

    retcode = MH_SetInputChannelEnable(dev[0], i, 1);
    if (retcode < 0)
    {
      MH_GetErrorString(Errorstring, retcode);
      printf("\nMH_SetInputChannelEnable error %d (%s). Aborted.\n", retcode, Errorstring);
      goto ex;
    }
  }

  if (Mode != MODE_T2)
  {
    retcode = MH_SetBinning(dev[0], Binning);
    if (retcode < 0)
    {
      MH_GetErrorString(Errorstring, retcode);
      printf("\nMH_SetBinning error %d (%s). Aborted.\n", retcode, Errorstring);
      goto ex;
    }

    retcode = MH_SetOffset(dev[0], Offset);
    if (retcode<0)
    {
      MH_GetErrorString(Errorstring, retcode);
      printf("\nMH_SetOffset error %d (%s). Aborted.\n", retcode, Errorstring);
      goto ex;
    }
  }

  retcode = MH_GetResolution(dev[0], &Resolution);
  if (retcode<0)
  {
    MH_GetErrorString(Errorstring, retcode);
    printf("\nMH_GetResolution error %d (%s). Aborted.\n", retcode, Errorstring);
    goto ex;
  }
  printf("\nResolution is %1.0lfps\n", Resolution);


  printf("\nMeasuring input rates...\n");


  // After Init allow 150 ms for valid  count rate readings
  // Subsequently you get new values after every 100ms
  Sleep(150);

  retcode = MH_GetSyncRate(dev[0], &Syncrate);
  if (retcode < 0)
  {
    MH_GetErrorString(Errorstring, retcode);
    printf("\nMH_GetSyncRate error%d (%s). Aborted.\n", retcode, Errorstring);
    goto ex;
  }
  printf("\nSyncrate=%1d/s", Syncrate);


  for (i = 0; i < NumChannels; i++) // for all channels
  {
    retcode = MH_GetCountRate(dev[0], i, &Countrate);
    if (retcode<0)
    {
      MH_GetErrorString(Errorstring, retcode);
      printf("\nMH_GetCountRate error %d (%s). Aborted.\n", retcode, Errorstring);
      goto ex;
    }
    printf("\nCountrate[%1d]=%1d/s", i, Countrate);
  }

  printf("\n");

  //after getting the count rates you can check for warnings
  retcode = MH_GetWarnings(dev[0], &warnings);
  if (retcode<0)
  {
   MH_GetErrorString(Errorstring, retcode);
    printf("\nMH_GetWarnings error %d (%s). Aborted.\n", retcode, Errorstring);
    goto ex;
  }
  if (warnings)
  {
    MH_GetWarningsText(dev[0], warningstext, warnings);
    printf("\n\n%s", warningstext);
  }


  printf("\npress RETURN to start");
  getchar();

  printf("\nStarting data collection...\n");

  Progress = 0;
  printf("\nProgress:%12u", Progress);

  retcode = MH_StartMeas(dev[0], Tacq);
  if (retcode < 0)
  {
    MH_GetErrorString(Errorstring, retcode);
    printf("\nMH_StartMeas error %d (%s). Aborted.\n", retcode, Errorstring);
    goto ex;
  }

  while (1)
  {
    retcode = MH_GetFlags(dev[0], &flags);
    if (retcode < 0)
    {
      MH_GetErrorString(Errorstring, retcode);
      printf("\nMH_GetFlags error %d (%s). Aborted.\n", retcode, Errorstring);
      goto ex;
    }

    if (flags & FLAG_FIFOFULL)
    {
      printf("\nFiFo Overrun!\n");
      goto stoptttr;
    }

    retcode = MH_ReadFiFo(dev[0], buffer, &nRecords);	//may return less!  
    if (retcode < 0)
    {
      MH_GetErrorString(Errorstring, retcode);
      printf("\nMH_ReadFiFo error %d (%s). Aborted.\n", retcode, Errorstring);
      goto stoptttr;
    }

    if (nRecords)
    {
      if (fwrite(buffer, 4, nRecords, fpout) != (unsigned)nRecords)
      {
        printf("\nfile write error\n");
        goto stoptttr;
      }
      Progress += nRecords;
      printf("\b\b\b\b\b\b\b\b\b\b\b\b%12u", Progress);
      fflush(stdout);
    }
    else
    {
      retcode = MH_CTCStatus(dev[0], &ctcstatus);
      if (retcode < 0)
      {
        MH_GetErrorString(Errorstring, retcode);
        printf("\nMH_CTCStatus error %d (%s). Aborted.\n", retcode, Errorstring);
        goto ex;
      }
      if (ctcstatus)
      {
        printf("\nDone\n");
        goto stoptttr;
      }
    }

    //within this loop you can also read the count rates if needed.
  }

stoptttr:

  retcode = MH_StopMeas(dev[0]);
  if (retcode < 0)
  {
   MH_GetErrorString(Errorstring, retcode);
    printf("\nMH_StopMeas error %d (%s). Aborted.\n", retcode, Errorstring);
    goto ex;
  }

ex:

  for (i = 0; i < MAXDEVNUM; i++) //no harm to close all
  {
    MH_CloseDevice(i);
  }
  if (fpout)
  {
    fclose(fpout);
  }

  printf("\npress RETURN to exit");
  getchar();

  return 0;
}

File: clib/demos/32bit/Matlab/tttrmode.m

% Demo for access to MultiHarp 150 Hardware via MHLIB.DLL v 2.0.
% The program performs a TTTR measurement based on hardcoded settings.
% The resulting data stream is stored in a binary output file.
%
% Axel Hagen, PicoQuant, May 2020


% Constants from mhdefin.h

REQLIBVER   =   '2.0';    % this is the version this program expects
MAXDEVNUM   =       8;
TTREADMAX   = 1048576;    % 1M event records 
MODE_HIST   =       0;
MODE_T2	   =        2;
MODE_T3	   =        3;

FLAG_FIFOFULL = hex2dec('0002');

TRGLVLMIN	  =      -1200;	    % mV
TRGLVLMAX	  =       1200;	    % mV
OFFSETMIN	  =          0;     % ns
OFFSETMAX	  =  100000000;	    % ns

ACQTMIN		  =          1;	    % ms
ACQTMAX		  =  360000000;	    % ms  (100*60*60*1000ms = 100h)

% Errorcodes from errorcodes.h

MH_ERROR_DEVICE_OPEN_FAIL		 = -1;

% Settings for the measurement, Adapt to your setup!
 
Mode          = MODE_T3; %  you can change this
SyncTrgEdge   =       0; %  you can change this
SyncTrgLevel  =    -100; %  you can change this
InputTrgEdge  =       0; %  you can change this
InputTrgLevel =    -100; %  you can change this
SyncDiv       =       1; %  you can change this (observe mode!)
Binning       =       0; %  you can change this
Tacq          =   10000; %  you can change this       
    

fprintf('\nMultiHarp 150 MHLib.DLL Demo Application             PicoQuant 2020\n');

if (~libisloaded('MHlib'))    
    %Attention: The header file name given below is case sensitive and must
    %be spelled exactly the same as the actual name on disk except the file 
    %extension. 
    %Wrong case will apparently do the load successfully but you will not
    %be able to access the library!
    %The alias is used to provide a fixed spelling for any further access via
    %calllib() etc, which is also case sensitive.
    
    %To load the right dll for OS and bitness we use the mexext command.
    %Therefor the next four lines, demonstrating the old manual way, have
    %been commented.
    
    %loadlibrary('Mhlib.dll',   'mhlib.h', 'alias', 'MHlib'); % Windows 32 bit
    %loadlibrary('MHLib64.dll', 'mhlib.h', 'alias', 'MHlib'); % Windows 64 bit 
    %loadlibrary('/usr/local/lib/mh150/mhlib.so', 'mhlib.h', 'alias', 'MHlib'); % Linux 32 bit       
    %loadlibrary('/usr/local/lib64/mh150/mhlib.so', 'mhlib.h', 'alias', 'MHlib'); % Linux 64 bit
    
    OS = mexext;
    if strcmp('mexw32', OS)
        DLL = 'MHLib.dll';                        % Windows 32 bit
    elseif strcmp('mexw64', OS)
        DLL = 'MHLib64.dll';                      % Windows 64 bit
    elseif strcmp('mexa32', OS)
        DLL = '/usr/local/lib/mh150/mhlib.so';    % Linux 32 bit
    elseif strcmp('mexa64', OS)
        DLL = '/usr/local/lib64/mh150/mhlib.so';  % Linux 64 bit
    else
        fprintf('\nNo supported OS\n');
        return;
    end
    loadlibrary(DLL, 'mhlib.h', 'alias', 'MHlib');
else
    fprintf('Note: MHlib was already loaded\n');
end

if (libisloaded('MHlib'))
    fprintf('MHlib opened successfully\n');
    %libfunctionsview MHlib; %use this to test for proper loading
else
    fprintf('Could not open MHlib\n');
    return;
end

    
LibVersion    = blanks(8); %enough length!
LibVersionPtr = libpointer('cstring', LibVersion);

[ret, LibVersion] = calllib('MHlib', 'MH_GetLibraryVersion', LibVersionPtr);
if (ret<0)
    fprintf('Error in GetLibVersion. Aborted.\n');
    err = MH_GETLIBVERSION_ERROR;
else
	fprintf('MHLib version is %s\n', LibVersion);
end

if ~strcmp(LibVersion,REQLIBVER)
    fprintf('This program requires MHLib version %s\n', REQLIBVER);
    return;
end

fid = fopen('tttrmode.out','wb');
if (fid<0)
    fprintf('Cannot open output file\n');
    return;
end

fprintf('Measurement Mode  : T%ld\n',Mode);
fprintf('Binning           : %ld\n',Binning);
fprintf('AcquisitionTime   : %ld\n',Tacq);
fprintf('SyncDivider       : %ld\n',SyncDiv);
fprintf('SyncTrgEdge       : %ld\n',SyncTrgEdge);
fprintf('SyncTrgLevel      : %ld\n',SyncTrgLevel);
fprintf('InputTrgEdge      : %ld\n',InputTrgEdge);
fprintf('InputTrgLevel     : %ld\n',InputTrgLevel);


fprintf('\nSearching for MultiHarp devices...');

dev = [];
found = 0;
Serial     = blanks(8); %enough length!
SerialPtr  = libpointer('cstring', Serial);
ErrorStr   = blanks(40); %enough length!
ErrorPtr   = libpointer('cstring', ErrorStr);

for i=0:MAXDEVNUM-1
    [ret, Serial] = calllib('MHlib', 'MH_OpenDevice', i, SerialPtr);
    if (ret==0)       % Grab any MultiHarp we successfully opened
        fprintf('\n  %1d        S/N %s', i, Serial);
        found = found+1;            
        dev(found)=i; %keep index to devices we may want to use
    else
        if(ret==MH_ERROR_DEVICE_OPEN_FAIL)
            fprintf('\n  %1d        no device', i);
        else 
            [ret, ErrorStr] = calllib('MHlib', 'MH_GetErrorString', ErrorPtr, ret);
            fprintf('\n  %1d        %s', i,ErrorStr);
        end
    end
end

    
% in this demo we will use the first MultiHarp device we found, i.e. dev(1)
% you could also check for a specific serial number, so that you always know 
% which physical device you are talking to.

if (found<1)
	fprintf('\nNo device available. Aborted.\n');
	return; 
end

fprintf('\nUsing device #%1d',dev(1));
fprintf('\nInitializing the device...');

[ret] = calllib('MHlib', 'MH_Initialize', dev(1), Mode, 0);
if(ret<0)
	fprintf('\nMH_Initialize error %d. Aborted.\n',ret);
    closedev;
	return;
end 

%this is only for information
Model      = blanks(24); %enough length!
Partno     = blanks(8); %enough length!
Version    = blanks(8); %enough length!
ModelPtr   = libpointer('cstring', Model);
PartnoPtr  = libpointer('cstring', Partno);
VersionPtr = libpointer('cstring', Version);

[ret, Model, Partno] = calllib('MHlib', 'MH_GetHardwareInfo', dev(1), ModelPtr, PartnoPtr, VersionPtr);
if (ret<0)
    fprintf('\nMH_GetHardwareInfo error %1d. Aborted.\n',ret);
    closedev;
	return;
else
	fprintf('\nFound model %s part number %s version %s', Model, Partno, Version);             
end

NumInpChannels = int32(0);
NumChPtr = libpointer('int32Ptr', NumInpChannels);
[ret,NumInpChannels] = calllib('MHlib', 'MH_GetNumOfInputChannels', dev(1), NumChPtr); 
if (ret<0)
    fprintf('\nMH_GetNumOfInputChannels error %1d. Aborted.\n',ret);
    closedev;
	return;
else
	fprintf('\nDevice has %i input channels.', NumInpChannels);             
end
   
[ret] = calllib('MHlib', 'MH_SetSyncDiv', dev(1), SyncDiv);
if (ret<0)
    fprintf('\nMH_SetSyncDiv error %1d. Aborted.\n',ret);
    closedev;
    return;
end

[ret] = calllib('MHlib', 'MH_SetSyncEdgeTrg', dev(1), SyncTrgLevel, SyncTrgEdge);
if (ret<0)
    fprintf('\nMH_SetSyncEdgeTrg error %ld. Aborted.\n', ret);
    closedev;
    return;
end
 
[ret] = calllib('MHlib', 'MH_SetSyncChannelOffset', dev(1), 0);
if (ret<0)
   fprintf('\nMH_SetSyncChannelOffset error %ld. Aborted.\n', ret);
   closedev;
   return;
end 

for i=0:NumInpChannels-1 % we use the same input settings for all channels
    [ret] = calllib('MHlib', 'MH_SetInputEdgeTrg', dev(1), i, InputTrgLevel, InputTrgEdge);
    if (ret<0)
        fprintf('\nMH_SetInputEdgeTrg error %ld. Aborted.\n', ret);
        closedev;
        return;
    end
    [ret] = calllib('MHlib', 'MH_SetInputChannelOffset', dev(1), i, 0);
    if (ret<0)
        fprintf('\nMH_SetInputChannelOffset error %ld. Aborted.\n', ret);
        closedev;
        return;
    end   
end

if (Mode ~= MODE_T2)    %MH_SetBinning is meaningless in T2 mode
    [ret] = calllib('MHlib', 'MH_SetBinning', dev(1), Binning);
    if (ret<0)
        fprintf('\nMH_SetBinning error %ld. Aborted.\n', ret);
        closedev;
        return;
    end
end

if (Mode ~= MODE_T2)    %MH_SetOffset is meaningless in T2 mode
    [ret] = calllib('MHlib', 'MH_SetOffset', dev(1), 0);
    if (ret<0)
        fprintf('\nMH_SetOffset error %ld. Aborted.\n', ret);
        closedev;
        return;
    end
end

%MH_GetResolution is meaningless in T2 mode but it does no harm
Resolution = 0;
ResolutionPtr = libpointer('doublePtr', Resolution);
[ret, Resolution] = calllib('MHlib', 'MH_GetResolution', dev(1), ResolutionPtr);
if (ret<0)
    fprintf('\nMH_GetResolution error %ld. Aborted.\n', ret);
    closedev;
    return;
end
fprintf('\nResolution=%1dps', Resolution);


pause(0.4); % after Init or SetSyncDiv you must allow 400 ms for valid new count rates
            % otherwise you get new values after every 100 ms


% from here you can repeat the measurement (with the same settings)


Syncrate = 0;
SyncratePtr = libpointer('int32Ptr', Syncrate);
[ret, Syncrate] = calllib('MHlib', 'MH_GetSyncRate', dev(1), SyncratePtr);
if (ret<0)
    fprintf('\nMH_GetSyncRate error %ld. Aborted.\n', ret);
    closedev;
    return;
end
fprintf('\nSyncrate=%1d/s', Syncrate);
 
for i=0:NumInpChannels-1
	Countrate = 0;
	CountratePtr = libpointer('int32Ptr', Countrate);
	[ret, Countrate] = calllib('MHlib', 'MH_GetCountRate', dev(1), i, CountratePtr);
	if (ret<0)
        fprintf('\nMH_GetCountRate error %ld. Aborted.\n', ret);
        closedev;
        return;
    end
	fprintf('\nCountrate%1d=%1d/s', i, Countrate);
end


Warnings = 0;
WarningsPtr = libpointer('int32Ptr', Warnings);
[ret, Warnings] = calllib('MHlib', 'MH_GetWarnings', dev(1), WarningsPtr);
if (ret<0)
    fprintf('\nMH_GetWarnings error %ld. Aborted.\n',ret);
    closedev;
    return;
end
if (Warnings~=0)
    Warningstext = blanks(16384); %enough length!
    WtextPtr     = libpointer('cstring', Warningstext);
    [~, Warningstext] = calllib('MHlib', 'MH_GetWarningsText', dev(1), WtextPtr, Warnings);
    fprintf('\n\n%s',Warningstext);
end


buffer  = uint32(zeros(1,TTREADMAX));
bufferptr = libpointer('uint32Ptr', buffer);

nactual = int32(0);
nactualptr = libpointer('int32Ptr', nactual);

ctcdone = int32(0);
ctcdonePtr = libpointer('int32Ptr', ctcdone);

Progress = 0;
fprintf('\nProgress:%9d',Progress);
       
ret = calllib('MHlib', 'MH_StartMeas', dev(1),Tacq); 
if (ret<0)
    fprintf('\nMH_StartMeas error %ld. Aborted.\n', ret);
    closedev;
    return;
end
       
while(1)  
    
    flags = int32(0);
    flagsPtr = libpointer('int32Ptr', flags);
    [ret,flags] = calllib('MHlib', 'MH_GetFlags', dev(1), flagsPtr);
    if (ret<0)
    	  fprintf('\nMH_GetFlags error %ld. Aborted.\n', ret);
    	  break
    end
   
    if (bitand(uint32(flags),FLAG_FIFOFULL)) 
        fprintf('\nFiFo Overrun!\n'); 
        break;
    end
		
    [ret, buffer, nactual] = calllib('MHlib','MH_ReadFiFo', dev(1), bufferptr, nactualptr);
    %Note that MH_ReadFiFo may return less than requested  
    if (ret<0)  
        fprintf('\nMH_ReadFiFo error %d\n',ret); 
        break;
    end  

    if(nactual) 
        cnt = fwrite(fid, buffer(1:nactual),'uint32');
        if(cnt ~= nactual)
            fprintf('\nfile write error\n');
            break;
        end          
		Progress = Progress + nactual;
		fprintf('\b\b\b\b\b\b\b\b\b%9d',Progress);
    else
        [ret,ctcdone] = calllib('MHlib', 'MH_CTCStatus', dev(1), ctcdonePtr);
        if (ret<0)  
            fprintf('\nMH_CTCStatus error %d\n',ret); 
            break;
        end       
        if (ctcdone) 
            fprintf('\nDone\n'); 
            break;
        end
    end
     
	 %you can read the count rates here if needed

end %while


ret = calllib('MHlib', 'MH_StopMeas', dev(1)); 
if (ret<0)
    fprintf('\nMH_StopMeas error %ld. Aborted.\n', ret);
    closedev;
    return;
end
        
closedev;
    
fprintf('\nBinary output data is in tttrmode.out\n');

if(fid>0) 
    fclose(fid);
end

